internal/operands/common-templates/reconcile.go

```diff
@@ -122,6 +127,55 @@ func reconcileEditRole(request *common.Request) (common.ResourceStatus, error) {
 			foundRole.Rules = newRole.Rules
 		})
 }
+func updateOldTemplates(request *common.Request) (common.ResourceStatus, error) {
+	baseRequirement, err := labels.NewRequirement("template.kubevirt.io/type", selection.Equals, []string{"base"})
+	if err != nil {
+		request.Logger.Error(err, fmt.Sprintf("Error during preparing of type label requirement, %v", err))
+		panic(err)
+	}
+	var versionRequirement *labels.Requirement
+	versionRequirement, err = labels.NewRequirement("template.kubevirt.io/version", selection.NotEquals, []string{Version})
+	if err != nil {
+		request.Logger.Error(err, fmt.Sprintf("Error during preparing of version label requirement, %v", err))
+		panic(err)
+	}
+
+	labelsSelector := labels.NewSelector().Add(*baseRequirement, *versionRequirement)
+
+	opts := client.ListOptions{
+		LabelSelector: labelsSelector,
+	}
+
+	var oldTemplates templatev1.TemplateList
+	err = request.Client.List(request.Context, &oldTemplates, &opts)
+	if err != nil {
+		request.Logger.Error(err, fmt.Sprintf("Error during listing templates, %v", err))
+		return common.ResourceStatus{}, err
+	}
+
+	updatedTemplates := make([]templatev1.Template, 0)
+	for _, template := range oldTemplates.Items {
+		updated := false
+		for key := range template.Labels {
+			if strings.HasPrefix(key, "os.template.kubevirt.io/") || strings.HasPrefix(key, "flavor.template.kubevirt.io/") || strings.HasPrefix(key, "workload.template.kubevirt.io/") {
+				updated = true
+				delete(template.Labels, key)
+			}
+		}
+		if updated {
+			updatedTemplates = append(updatedTemplates, template)
+		}
+	}
+	oldTemplates.Items = updatedTemplates
+	for _, template := range oldTemplates.Items {
+		err = request.Client.Update(request.Context, &template)
```

Comment:
I think instead of Updates I'd do Patches to the list of resources.
You can build a JSON Patch from the fields that should be removed:
```
[ { "op": "remove", "path": "/metadata/labels/os.template.kubevirt.io\/centos8" }]
```
So I'd collect all the labels that should be removed for each resource (or just all of them) and then apply the patch.
This should be a bit less error prone as it does not care about the previous object. I think Update "could" fail if there was a modification since the `get`.